src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,344 @@
+            _tryTakeTimeout = tryTakeTimeout;
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _removeExpiredValues = removeExpiredValues;
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            if (_removeExpiredValues)
+                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, _tryTakeTimeout))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }          
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (true)
+            {
+                if (_eventQueue.Count == 0)
+                    return;
+            }
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        internal void RemoveLRUs()
+        {
+            // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
+            int currentCount = _map.Count <= _capacity ? _capacity : _map.Count;
+            var newCacheSize = currentCount - (int)(currentCount * _compactionPercentage);
+            while (_map.Count > newCacheSize && _doubleLinkedList.Count > 0)
+            {
+                var lru = _doubleLinkedList.Last;
+                _map.TryRemove(lru.Value.Key, out _);
+                _doubleLinkedList.Remove(lru);
+            }
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (!_disposed)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
```

Comment:
>_eventQueue [](start = 16, length = 11)

there is a theoretical probability that this could be disposed... perhaps we should have try catch around it?